WIP - Google Calendar Background Worker

NickyBobby opened the pull request.

NickyBobby committed:
Starts background worker implementation

jedschneider committed:
wip on scheduling

jedschneider committed:
scheduler works, I think

jedschneider committed:
adds some other use cases

NickyBobby committed:
Adds background worker to fetch google calendars

* Adds sidekiq-cron, whenever, and redis gems
* I set up a cron job for a a background worker to run every morning at 6 am
* The worker then sets up jobs, which is the google calendar fetch for each student

Still need to do:

* Have each cron job set up the next cron job - 15 minute intervals
* Figure out how to set up redis and also how to test it
* Set up a presenter from the google calendar object to a hash for the CalendarZipper

NickyBobby committed:
Adds tests for checksum helper

erinnachen commented on app/controllers/callbacks/google_controller.rb:
```diff
@@ -0,0 +1,9 @@
+
+  # This was supposed to be the endpoint to set up push notifications
+  # It is not operational right now
+  def watch
```
one thing I realized is that you may need to have an authorization on this

erinnachen commented on app/workers/update_schedule_worker.rb:
```diff
@@ -0,0 +1,10 @@
+  def self.perform
+    Student.all.each do |student|
+      student.update(at_school: false)
+      UpdateScheduleJob.perform(student)
```
I think the default at_school should be `true`

NickyBobby replied:
Well I was thinking that when they sign in it changes the value to true, then every morning at 6 we set it to false. Are you saying we should set it to true every morning? @erinnachen

erinnachen replied:
Yup. Assuming that more students are usually present than absent, then you'll have to manually adjust fewer people. Also at the beginning of the day, you'll see those who have not checked in as lost (which is at the top of the grove monitor) instead of absent which is at the bottom of the page.

NickyBobby replied:
So then how would they be marked absent? If they don't check in to their first class, would it go to false?

erinnachen replied:
then they would be in the lost bin and teachers can then go and manually mark them absent.

NickyBobby committed:
Adds Google Service test and Calendar Zipper parser

NickyBobby committed:
Adds ability to set schedule to Redis and tests for it

NickyBobby committed:
Adds functionality for each job to schedule the next job

NickyBobby committed:
Adds initial setup for push notifications for google calendars

NickyBobby commented:
This PR is up to date and rebased over the current master. Things that still need to be done
- mock out google api tests
- Write tests for push notifications callback controller
- Create Event objects for the `UpdateScheduleJob`
- finish off notifications `watch` action to activate `UpdateScheduleJob`

NickyBobby committed:
starts google controller callback functionality

jedschneider commented:
## cleanup
- [x] remove pry statements
- [x] add cassettes to project
- [x] remove checksum stuff
- [x] remove hash_key from student model
- [x] site verification html file in public?
- [x] clean up any private info in the rails helper

## stuff to add
- [x] event metadata from google to event
- [x] setting event properties from google event
- [x] store the activity id on the the scan record so we know were to restart the playlist
- [x] store the start of the playlist for the day on the user model via redis so we know where to restart the schedule from if the schedule during the day changes.
- [x] removing old events, creating new ones when the background job runs
- [ ] possibly a rake task to populate a student with a schedule (to make it easy to style)

NickyBobby committed:
Cleans up UpdateScheduleJob

- Switches schedule over from Redis key/value storage to Event model
- Adds more data to the CalendarEventParser
- Cleans up Rails Helper
- MIGRATION adds title and metadata to the Event model
- Removes sensitive information from the VCR cassettes

NickyBobby committed:
Adds unit test for CalendarEventParser

NickyBobby commented:
Hey @jedschneider I'm not sure why the CI build is still failing. I added the cassettes to the PR and they all pass locally. I do have one test being skipped right now, do you think that's the issue? Also I'm unit testing the CalendarEventParser. There is an issue with setting the creator_id that I'd like to talk to you about. Also I'm not sure which is the best way to add activity_id to each scan. Hit me up when you get a chance.

NickyBobby committed:
Adds activity_id to Scan model

NickyBobby committed:
Adds last playlist activity for the day on redis

NickyBobby committed:
Fixes travis CI build

NickyBobby committed:
calendar zipper still WIP

NickyBobby committed:
Fixes UpdateScheduleJob perform function

NickyBobby commented:
hey @jedschneider I was able to fix the calendar zipper and get all the tests to pass. I changed the activity attribute from name to title because I thought it would've been easier to pass in the activity attribute directly to the event creation. It broke a lot of tests and I fixed them all. In the end it was stupid because we still need to normalize the data that's going into event creation. So it works for right now but definitely not the most elegant solution to say the least...

jedschneider committed:
grab activity from scan not event

jedschneider committed:
get watch controller test working

NickyBobby committed:
Scopes google calendar search to within grove master calendar timeframe

NickyBobby committed:
Adds Schedule Normalizer to create events

NickyBobby committed:
Changes postgresql version on travis.yml file

jedschneider committed:
moving to a specific yml for travis

jedschneider committed:
run dangit

jedschneider committed:
adds redis to travis

NickyBobby committed:
Adds heroku scheduler rake task

jedschneider committed:
removes idea of periodicity from master calendar - now doing just start and end of day

jedschneider committed:
adds route to rebuild schedule on demand

jedschneider committed:
adds button to rebuild schedule for a student and enumerates the current schedule

jedschneider committed:
adds deploy script

NickyBobby commented:
Looks TOO good NOT to merge! 👍
